### PR TITLE
fix/NSS#1208

### PR DESCRIPTION
**Pull request by jaxoncreed**
Fix for https://github.com/solid/node-solid-server/issues/1208

### CODE REVIEW

File: src/SolidClient.js [hunk 1]
```diff
@@ -174,8 +174,29 @@ class SolidClient {
     }
 
     // Redirect to the authentication page, passing the session cookie
-    const authUrl = loginResponse.headers.location;
+    let authUrl = loginResponse.headers.location;
     const cookie = loginResponse.headers['set-cookie'][0].replace(/;.*/, '');
+
+    // Handle the new consent page in 5.1.1
+    if (this.isAboveVersion511(loginResponse.headers['x-powered-by'])) {
```
Comment (RubenVerborgh) [inline src/SolidClient.js:181]: Oh god, this is a horrible hack.
I take it there is no better way?

File: src/SolidClient.js [hunk 2]
```diff
@@ -174,8 +174,29 @@ class SolidClient {
...
+      const consUrl = new URL(authUrl);
```
Comment (RubenVerborgh) [inline src/SolidClient.js]: `consentUrl`?

File: src/SolidClient.js [hunk 3]
```diff
@@ -174,8 +174,29 @@ class SolidClient {
...
+      const search = consUrl.search.substring(1);
+      let consPostData = JSON.parse('{"' + decodeURI(search).replace(/"/g, '\\"').replace(/&/g, '","').replace(/\=/g, '":"') + '"}');
```
Comment (RubenVerborgh) [inline src/SolidClient.js]: Never use `decodeURI`; always use `decodeURIComponent`.

(Reason: `encodeURI('http://example.org/?abc') === 'http://example.org/?abc'`)

File: src/SolidClient.js [hunk 4]
```diff
@@ -185,6 +206,21 @@ class SolidClient {
...
+      const versionNumbers = versionNumber.split('.');
+      for (let i = 0; i < targetVersion.length; i++) {
+        if (targetVersion[i] > versionNumbers[i]) {
+          return false;
+        }
+      }
+      return true;
+    }
+    return false;
+  }
```
Comment (RubenVerborgh) [inline src/SolidClient.js:211]: I wouldn't over-engineer and just `return /^solid-server\/5\.(1\.[1-9]|[2-9]|1\d)/.test(version);` (assuming that there will never be a NSS v6).

**Review by RubenVerborgh — APPROVED** [review]
Thanks, total hack of course, but then again this entire library is.

File: src/SolidClient.js (hunk 3 above)
Comment (jaxoncreed) [reply src/SolidClient.js]: Haha yeah. I just noticed that myself.

File: src/SolidClient.js (hunk 1 above)
Comment (jaxoncreed) [reply src/SolidClient.js:181]: The better way would be re-writing this library to handle a more automated oidc flow haha

**Comment by jaxoncreed** [comment]
Okay this is now good to merge @RubenVerborgh

**Comment by jeff-zucker** [comment]
Just to confirm, I was able to get this from a test script with this branch:
```
logged in as <https://jeffzucker.dev.inrupt.net/profile/card#me>
Ok GET public resource from solid-server/5.1.3
Ok PUT resource in private container.
```

**Comment by RubenVerborgh** [comment]
Thanks @jaxoncreed and @jeff-zucker.

@jaxoncreed Do you want to become a maintainer of the package? Then I'd give you the right permissions for merging and npm publishing. If you prefer not to, I can follow up.

**Comment by jaxoncreed** [comment]
@RubenVerborgh Sure. I can take that on.

**Comment by jeff-zucker** [comment]
:+1:  It will be great to have someone on the server side managing this  :-).  Please count on me for any testing or other help you need.

**Comment by RubenVerborgh** [comment]
>  It will be great to have someone on the server side managing this :-)

Hey hey, I've done server work 😛 But not lately indeed.

> Please count on me for any testing or other help you need.

Excellent, thanks!

**Comment by jeff-zucker** [comment]
> I've done server work

Ha, yes, I'm sure you have.  I just meant that  given such strong (and in the long run inappropriate)  binding of app and server, it will be good to have the catching designed by someone who does the pitching :-).

**Comment by jaxoncreed** [comment]
@RubenVerborgh Could I also have npm access to this? My username is the same (jaxoncreed)

**Comment by jeff-zucker** [comment]
@jaxoncreed, if you npm it, please include https://github.com/solid/solid-cli/pull/5, otherwise rdflib will continue to show vulnerabilities.

**Comment by RubenVerborgh** [comment]
@jaxoncreed Best to have access to the entire `@solid` group in npm; Inrupt can help you with that.

**Comment by jaxoncreed** [comment]
@jeff-zucker absolutely. Good catch

**Pull request merged**